dvc/remote/local/__init__.py

```diff
@@ -564,5 +589,97 @@ def unprotect(path_info):
             RemoteLOCAL._unprotect_file(path)
 
     @staticmethod
-    def protect(path_info):
-        os.chmod(path_info.path, stat.S_IREAD | stat.S_IRGRP | stat.S_IROTH)
+    def protect(path):
+        os.chmod(path, stat.S_IREAD | stat.S_IRGRP | stat.S_IROTH)
+
+    def append_verififcation_suffix(self, string):
+        return string + "." + self.STATUS_VERIFICATION_DIR_SUFFIX
+
+    def get_verification_dir_info(self, checksum):
+        verification_dir_info = self.checksum_to_path_info(checksum)
+        verification_dir_info.path = self.append_verififcation_suffix(
+            verification_dir_info.path
+        )
+
+        return verification_dir_info
+
+    def update_verification_dir(self, checksum, dir_info):
+        verification_dir_info = self.get_verification_dir_info(checksum)
+
+        exists = self.exists(verification_dir_info)
+        if exists:
+            if self._path_metadata_up_to_date(verification_dir_info.path):
+                return
+
+            shutil.rmtree(verification_dir_info.path)
+
+        self._create_verification_dir(
+            checksum, dir_info, verification_dir_info
+        )
+
+    def _create_verification_dir(
+        self, checksum, dir_info, verification_dir_info
+    ):
+        self.makedirs(verification_dir_info)
+        for entry in dir_info:
+            entry_cache_info = self.checksum_to_path_info(
+                entry[self.PARAM_CHECKSUM]
+            )
+            relpath = entry[self.PARAM_RELPATH]
+
+            verification_entry_info = copy(entry_cache_info)
+            verification_entry_info.path = self.ospath.join(
+                verification_dir_info.path, relpath
+            )
+            self.link(entry_cache_info, verification_entry_info, "hardlink")
+        self.state.save(verification_dir_info, checksum)
+
+    def _path_metadata_up_to_date(self, path):
+        mtime, size = get_mtime_and_size(path)
+        inode = get_inode(path)
+
+        entry = self.state.get_state_record_for_inode(inode)
+
+        if entry:
+            state_mtime, state_size, _, _ = entry
+            if state_mtime == mtime and state_size == size:
+                return True
+        return False
+
+    def _is_verification_dir_up_to_date(self, checksum):
+        status_verification_dir_info = self.get_verification_dir_info(checksum)
+
+        if os.path.exists(status_verification_dir_info.path):
+            return self._path_metadata_up_to_date(
+                status_verification_dir_info.path
+            )
+        return False
+
+    def _changed_dir_cache(self, checksum):
+        if self._is_verification_dir_up_to_date(checksum):
```

Comment:
If upacked dir is not up to date, then we should fallback to manual check, don't we? Because in this current form you might get false-negatives.

Reply:
It looks like currently if unpacked dir is up to date you go ahead and check it manually once again. Seems strange. Maybe I'm missing something.

Reply:
I don't understand: 
- if verification_dir is up to date, changed_cache will return False, isn't it the point?

Reply:
Sorry, that is just me being dumb.

Reply:
After discussion:
logic was ok, but it was counterintuitive. Code changed so the question ask is always 
"did something change"  and not mixed with "is up to date"